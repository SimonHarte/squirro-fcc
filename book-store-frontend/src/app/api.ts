import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';
import endpoints from '../config/endpoints.json';

export const API = createApi({
  reducerPath: 'api',
  baseQuery: fetchBaseQuery({
    baseUrl: `${endpoints.api.url}:${endpoints.api.port}`,
  }),
  endpoints: (build) => ({
    getStoresData: build.query({
      query: () => `stores`,
    }),
    getBooksData: build.query({
      query: () => `books`,
    }),
    getAuthorsData: build.query({
      query: () => `authors`,
    }),
    getCountryById: build.query({
      query: (id: string) => `countries/${id}`,
    }),
    setNewRating: build.mutation({
      query: ({ id, rating }) => {
        const patch = {
          data: {
            id,
            type: 'stores',
            attributes: { rating },
          },
        };

        return {
          url: `stores/${id}`,
          method: 'PATCH',
          body: patch,
        };
      },
    }),
  }),
});

export const {
  useGetStoresDataQuery,
  useGetBooksDataQuery,
  useGetAuthorsDataQuery,
  useGetCountryByIdQuery,
  useSetNewRatingMutation,
} = API;
