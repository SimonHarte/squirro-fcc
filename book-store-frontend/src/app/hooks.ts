import type { Author } from '../types/Authors';
import type { Book, BestSeller } from '../types/Books';
import type { DataEntry } from '../types/Base';
import {
  useGetBooksDataQuery,
  useGetAuthorsDataQuery,
  useGetCountryByIdQuery,
} from './api';

export const useGetCountryCodeById = (id: string): string | null => {
  const { data: country, isLoading } = useGetCountryByIdQuery(id);

  return isLoading ? null : country.data.attributes.code;
};

export const useGetBestsellers = (books: Array<string>): Array<BestSeller> => {
  const { data: booksData, isLoading: isLoadingBooks } =
    useGetBooksDataQuery('');
  const { data: authorsData, isLoading: isLoadingAuthors } =
    useGetAuthorsDataQuery('');

  if (isLoadingBooks || isLoadingAuthors) {
    return [];
  }

  const bestsellers = booksData.data
    .filter((book: Book) => {
      return books?.includes(book.id);
    })
    .sort((prev: Book, next: Book) => {
      return next.attributes.copiesSold - prev.attributes.copiesSold;
    })
    .splice(0, 2)
    .map(
      ({
        attributes: { name },
        relationships: {
          author: {
            data: { id: authorId },
          },
        },
      }: Book): BestSeller => {
        const author: Author = authorsData.data.find(
          ({ id }: DataEntry) => id === authorId
        );

        return {
          name,
          author: author.attributes.fullName,
        };
      }
    );

  return bestsellers;
};
