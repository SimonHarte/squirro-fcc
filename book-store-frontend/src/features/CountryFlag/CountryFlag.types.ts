export type CountryFlagProps = {
  id: string;
};
