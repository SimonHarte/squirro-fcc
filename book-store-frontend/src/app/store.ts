import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import { storesApi } from './api';

export const store = configureStore({
  reducer: {
    [storesApi.reducerPath]: storesApi.reducer,
  },
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware().concat(storesApi.middleware);
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
