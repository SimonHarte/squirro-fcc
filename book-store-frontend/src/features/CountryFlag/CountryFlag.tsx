import { type ReactElement, type FC } from 'react';
import endpoints from '../../config/endpoints.json';
import { useGetCountryCodeById } from '../../app/hooks';
import type { CountryFlagProps } from './CountryFlag.types';
import styles from './CountryFlag.module.scss';

const CountryFlag: FC<CountryFlagProps> = ({ id }): ReactElement | null => {
  const code = useGetCountryCodeById(id);

  return code ? (
    <img
      className={styles.flag}
      src={`${endpoints.countryFlags.url}${code}`}
      alt={code}
    />
  ) : null;
};

export default CountryFlag;
