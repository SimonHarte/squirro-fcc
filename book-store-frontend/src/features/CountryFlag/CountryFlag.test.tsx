import { shallow } from 'enzyme';
import { useGetCountryCodeById } from '../../app/hooks';
import CountryFlag from './index';

jest.mock('../../app/hooks', () => ({
  useGetCountryCodeById: jest.fn(),
}));

const mockUseGetCountryCodeById = useGetCountryCodeById as jest.MockedFunction<
  typeof useGetCountryCodeById
>;

describe('CountryFlag', () => {
  test('should render correctly', () => {
    mockUseGetCountryCodeById.mockReturnValue('CH');

    const wrapper = shallow(<CountryFlag id="1" />);

    expect(wrapper).toMatchSnapshot();
  });

  test('should render correctly without code', () => {
    mockUseGetCountryCodeById.mockReturnValue(null);

    const wrapper = shallow(<CountryFlag id="9999" />);

    expect(wrapper).toMatchSnapshot();
  });
});
