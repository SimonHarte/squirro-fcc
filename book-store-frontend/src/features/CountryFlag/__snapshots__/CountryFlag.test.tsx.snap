// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryFlag should render correctly 1`] = `
<img
  alt="CH"
  className="flag"
  src="https://countryflagsapi.com/png/CH"
/>
`;

exports[`CountryFlag should render correctly without code 1`] = `""`;
